### PR TITLE
Avoid top-level `with` in `pkgs/development/compilers/gcc`

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

Using `with` at a top-level scope is [an anti-pattern](https://nix.dev/guides/best-practices#with-scopes). The tracking issue is #208242. This is a pure refactor: there is no functional change contained in this PR, though spot the bug fix where `fetchFromGitHub` wasn't injected!

I used the refactor strategy that looks for the uses of names [coming from `lib`](https://gist.github.com/philiptaron/04326000d5ff20cb72c2805efd09701c). 

Following advice, I also preferred to inherit names from `lib` instead of `builtins` where there was a choice. I also preferred to inherit from `lib` in one place, rather than to have a mixture of `lib.thing` and `inherit (lib) thing`.

For the assertions, I moved them into one block.

I really wanted to fix the indentation, but I didn't.

This PR targets staging, rather than master, as the contents of these files near the root of most of the Nixpkgs graph, but I guess I should have targeted master since there are no rebuilds.

See also:
- https://github.com/NixOS/nixpkgs/pull/297289

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/gcc/default.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, targetPackages, fetchurl, fetchpatch, noSysDirs
+{ lib, stdenv, targetPackages, fetchurl, fetchpatch, fetchFromGitHub, noSysDirs
```
Comment (philiptaron) [inline pkgs/development/compilers/gcc/default.nix:1]: The failure to accept `fetchFromGithub` was masked by the global `with`! This is a real bug.

File: pkgs/development/compilers/gcc/default.nix [hunk 2]
```diff
@@ -48,19 +48,37 @@
 }:
 
 let
-  versions = import ./versions.nix;
```
Comment (philiptaron) [inline pkgs/development/compilers/gcc/default.nix:51]: renamed to `gccVersions` so as not to conflict with `lib.versions` which was also used in this file.

File: pkgs/development/compilers/gcc/default.nix [hunk 3]
```diff
@@ -69,37 +87,9 @@ let
...
-# If turned on, we can't provide reproducible builds anymore
-assert reproducibleBuild -> profiledCompiler == false;
-
-with lib;
-with builtins;
+  is49 = majorVersion == "4" && versions.minor version == "9";
+  is48 = majorVersion == "4" && versions.minor version == "8";
 
-let inherit version;
     disableBootstrap = atLeast11 && !stdenv.hostPlatform.isDarwin && (atLeast12 -> !profiledCompiler);
```
Comment (philiptaron) [inline pkgs/development/compilers/gcc/default.nix:93]: Here's the indent bugaboo.

**Review by fricklerhandwerk — APPROVED** [review]
Great!

**Comment by philiptaron** [comment]
Ping for merge. 🙂

**Comment by fricklerhandwerk** [comment]
No one else checked it in the meantime, but 2 weeks should be enough time. I'll merge it, and anyone is free to revert if they disagree, no hard feelings.

**Pull request merged**